library/objective-c/EnvoyEngineImpl.m

```diff
@@ -10,6 +11,93 @@ static void ios_on_exit() {
   NSLog(@"[Envoy] library is exiting");
 }
 
+typedef struct {
+  __unsafe_unretained EnvoyHTTPFilter *filter;
+} ios_http_filter_context;
+
+// TODO(goaway): refactor and relocate translation
+static envoy_data toManagedNativeString(NSString *s) {
+  size_t length = s.length;
+  uint8_t *native_string = (uint8_t *)safe_malloc(sizeof(uint8_t) * length);
+  memcpy(native_string, s.UTF8String, length);
+  envoy_data ret = {length, native_string, free, native_string};
+  return ret;
+}
+
+static EnvoyHeaders *to_ios_headers(envoy_headers headers) {
+  NSMutableDictionary *headerDict = [NSMutableDictionary new];
+  for (envoy_header_size_t i = 0; i < headers.length; i++) {
+    envoy_header header = headers.headers[i];
+    NSString *headerKey = [[NSString alloc] initWithBytes:header.key.bytes
+                                                   length:header.key.length
+                                                 encoding:NSUTF8StringEncoding];
+    NSString *headerValue = [[NSString alloc] initWithBytes:header.value.bytes
+                                                     length:header.value.length
+                                                   encoding:NSUTF8StringEncoding];
+    NSMutableArray *headerValueList = headerDict[headerKey];
+    if (headerValueList == nil) {
+      headerValueList = [NSMutableArray new];
+      headerDict[headerKey] = headerValueList;
+    }
+    [headerValueList addObject:headerValue];
+  }
+  // TODO(goaway): consider solution that doesn't violate release convention
+  // Note: We don't call release_envoy_headers because they may not be modified by the filter
+  return headerDict;
+}
+
+static envoy_headers toNativeHeaders(EnvoyHeaders *headers) {
+  envoy_header_size_t length = 0;
+  for (id headerKey in headers) {
+    length += [headers[headerKey] count];
+  }
+  envoy_header *header_array = (envoy_header *)safe_malloc(sizeof(envoy_header) * length);
+  envoy_header_size_t header_index = 0;
+  for (id headerKey in headers) {
+    NSArray *headerList = headers[headerKey];
+    for (id headerValue in headerList) {
+      envoy_header new_header = {toManagedNativeString(headerKey),
+                                 toManagedNativeString(headerValue)};
+      header_array[header_index++] = new_header;
+    }
+  }
+  // TODO: ASSERT(header_index == length);
```

Comment:
We haven't actually (at present) introduced a dependency on Envoy's assertion library at this layer.

Reply:
Got it. If we want, we could use the native `assert`, but no strong opinion for now I guess

Reply:
For sure, we definitely can - we just haven't decided on a convention for this layer yet.

Reply:
Note a native `assert` won't get compiled out by Envoy's build directives, so it will always crash the app if we hit it. Not saying it's the right or wrong thing to do, but it's just something to consider.